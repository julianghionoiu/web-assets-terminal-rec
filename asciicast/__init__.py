import json
from decimal import Decimal

from colorama import Fore, Style


class Text:
    NEWLINE = "\r\n"
    CONSOLE_ARROW = Fore.CYAN + "âžœ " + Style.RESET_ALL


def wait(ticks):
    return [{"duration_in_ticks": ticks, "content": ""}]


def appear(content):
    return [{"duration_in_ticks": 1, "content": content}]


def typing(text="hostname"):
    return [{"duration_in_ticks": 1, "content": char} for char in text]


def delete(num_chars):
    return [{"duration_in_ticks": 1, "content": "\b \b"} for _ in range(0, num_chars)]


def press_enter():
    return [{"duration_in_ticks": 1, "content": Text.NEWLINE}]


def red(text):
    return Fore.RED + text + Style.RESET_ALL


def green(text):
    return Fore.GREEN + text + Style.RESET_ALL


def dim(text):
    return Style.DIM + text + Style.RESET_ALL


def bright(text):
    return Style.BRIGHT + text + Style.RESET_ALL


class DecimalEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Decimal):
            return float("{:.2f}".format(obj))
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)


def write_asciicast_v2(stream, metadata, asciicast_lines):
    length_of_one_tick_in_seconds = 0.10
    current_time_sec = Decimal(0.00)
    json.dump(metadata, stream)
    stream.write("\n")
    for line in asciicast_lines:
        asciicast_v2_line = [current_time_sec, "o", line["content"]]
        current_time_sec += Decimal(length_of_one_tick_in_seconds * line["duration_in_ticks"])

        json.dump(asciicast_v2_line, stream, cls=DecimalEncoder)
        stream.write("\n")
