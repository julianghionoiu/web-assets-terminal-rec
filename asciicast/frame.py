class AsciicastFrame(object):
    duration_in_ticks: int
    content: str

    def __init__(self, duration_in_ticks: int, content: str) -> None:
        self.duration_in_ticks = duration_in_ticks
        self.content = content
