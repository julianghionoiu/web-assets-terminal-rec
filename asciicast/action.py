from abc import ABC, abstractmethod
from typing import List

from asciicast.frame import AsciicastFrame


class AsciicastAction(ABC):

    @abstractmethod
    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        ...
