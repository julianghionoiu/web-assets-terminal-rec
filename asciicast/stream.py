import json
from decimal import Decimal

from asciicast.frame import AsciicastFrame


class AsciicastStream(object):
    width: int
    height: int

    def __init__(self, width: int, height: int, stream, length_of_one_tick_in_seconds=0.10) -> None:
        self.width = width
        self.height = height
        self._length_of_one_tick_in_seconds = length_of_one_tick_in_seconds

        # Start the stream
        self._stream = stream
        self._current_time_sec = Decimal(0.00)
        metadata = {"version": 2, "width": self.width, "height": self.height}
        json.dump(metadata, stream)
        stream.write("\n")

    def write_frame(self, frame: AsciicastFrame):
        asciicast_v2_line = [self._current_time_sec, frame.frame_type, frame.content]
        self._current_time_sec += Decimal(self._length_of_one_tick_in_seconds * frame.duration_in_ticks)
        json.dump(asciicast_v2_line, self._stream, cls=DecimalEncoder)
        self._stream.write("\n")


class DecimalEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Decimal):
            return float("{:.2f}".format(obj))
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)
