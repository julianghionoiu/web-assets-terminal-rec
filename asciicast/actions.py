from typing import List

from asciicast.action import AsciicastAction
from asciicast.frame import AsciicastFrame
from asciicast.text import newline


class Wait(AsciicastAction):
    _ticks: int

    def __init__(self, ticks: int) -> None:
        self._ticks = ticks

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(self._ticks, "")]


class Appear(AsciicastAction):
    _content: str

    def __init__(self, content: str) -> None:
        self._content = content

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, self._content)]


class TypeChars(AsciicastAction):
    _text: str

    def __init__(self, text: str) -> None:
        self._text = text

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, char) for char in self._text]


class Delete(AsciicastAction):
    _num_chars: int

    def __init__(self, num_chars: int) -> None:
        self._num_chars = num_chars

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, "\b \b") for _ in range(0, self._num_chars)]


class PressEnter(AsciicastAction):

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, newline())]
