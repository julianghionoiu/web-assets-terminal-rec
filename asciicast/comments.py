from asciicast.frame import AsciicastFrame, AsciicastFrameType
from asciicast.stream import AsciicastStream


class Comments(object):

    def __init__(self, asciicast_stream: AsciicastStream) -> None:
        self._asciicast_stream = asciicast_stream

    def add(self, comment: str) -> None:
        self._asciicast_stream.write_frame(
            AsciicastFrame(0,
                           frame_type=AsciicastFrameType.INPUT,
                           content=comment))
