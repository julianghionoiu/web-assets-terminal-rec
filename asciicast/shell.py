from asciicast.stream import AsciicastStream
from asciicast.text import newline


class Shell(object):

    def __init__(self, asciicast_stream: AsciicastStream) -> None:
        self._asciicast_stream = asciicast_stream

    def wait(self, ticks: int) -> None:
        self._asciicast_stream.write_frame(ticks, content="")

    def appear(self, content: str) -> None:
        self._asciicast_stream.write_frame(1, content=content)

    def type_chars(self, text: str) -> None:
        for _char in text:
            self._asciicast_stream.write_frame(1, content=_char)

    def delete(self, num_chars: int) -> None:
        for _ in range(0, num_chars):
            self._asciicast_stream.write_frame(1, content="\b \b")

    def press_enter(self):
        self._asciicast_stream.write_frame(1, content=newline())
