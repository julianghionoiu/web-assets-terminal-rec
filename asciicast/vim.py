
from stream import AsciicastStream


class VimEditor(object):
    _asciicast_stream: AsciicastStream
    _editor_contents: str

    def __init__(self, asciicast_stream: AsciicastStream, ) -> None:
        self._asciicast_stream = asciicast_stream
        self._editor_contents = ""
        self._cursor_row = 0
        self._cursor_column = 0
        self._viewport_row = 0
        self._viewport_column = 0

    def display_content(self, block_of_text):
        self._editor_contents = block_of_text

