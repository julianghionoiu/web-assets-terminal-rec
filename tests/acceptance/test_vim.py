import io

from approvaltests.approvals import verify

from asciicast.stream import AsciicastStream
from asciicast.vim import VimEditor


def test_cursor_move_down_then_up():
    string_stream = io.StringIO()
    asciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    vim = VimEditor(asciicast_stream, _generate_lines(25))

    vim.display_content()
    vim.cursor_down(num_rows=5)
    vim.cursor_down(num_rows=30)
    vim.cursor_up(num_rows=5)
    vim.cursor_up(num_rows=30)

    verify(string_stream.getvalue())


def test_cursor_move_right_then_left():
    string_stream = io.StringIO()
    asciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    vim = VimEditor(asciicast_stream, "123456789\n")

    vim.display_content()
    vim.cursor_right(num_cols=2)
    vim.cursor_right(num_cols=10)
    vim.cursor_left(num_cols=2)
    vim.cursor_left(num_cols=10)

    verify(string_stream.getvalue())


def test_cursor_move_right_with_colours():
    string_stream = io.StringIO()
    asciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    vim = VimEditor(asciicast_stream, "\u001b[31mredtext\u001b[0m\n")

    vim.display_content()
    vim.cursor_right(num_cols=2)
    vim.cursor_right(num_cols=10)

    verify(string_stream.getvalue())


def test_cursor_adjust_location_to_content():
    string_stream = io.StringIO()
    asciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    vim = VimEditor(asciicast_stream, "123456789\n12345\n12")

    vim.display_content()
    vim.cursor_right(num_cols=10)
    vim.cursor_down(num_rows=1)
    vim.content_scroll_down(1)

    verify(string_stream.getvalue())


def test_content_scroll_down_then_up():
    string_stream = io.StringIO()
    asciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    vim = VimEditor(asciicast_stream, _generate_lines(25))

    vim.display_content()
    vim.cursor_down(num_rows=2)
    vim.content_scroll_down(num_rows=2)
    vim.content_scroll_down(num_rows=20)
    vim.content_scroll_up(num_rows=2)
    vim.content_scroll_up(num_rows=20)

    verify(string_stream.getvalue())


def test_footer_appear_and_type():
    string_stream = io.StringIO()
    asciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    vim = VimEditor(asciicast_stream, "some_text\n")

    vim.display_content()
    vim.appear_in_footer("-- INSERT --")
    vim.type_in_footer(":wq")

    verify(string_stream.getvalue())


def test_content_delete_and_insert():
    string_stream = io.StringIO()
    asciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    vim = VimEditor(asciicast_stream, "\u001b[31mred????\u001b[0m word2\n")

    vim.display_content()
    vim.cursor_right(num_cols=3)
    vim.delete_at_cursor(num_chars=20)
    vim.type_at_cursor("word\n")

    verify(string_stream.getvalue())


def test_quitting():
    string_stream = io.StringIO()
    asciicast_stream = AsciicastStream(width=60, height=21, stream=string_stream)
    vim = VimEditor(asciicast_stream, "some content\n")

    vim.display_content()
    vim.close()

    verify(string_stream.getvalue())


# ~~~~~~ Test helper

def _generate_lines(num_lines):
    block_of_text: str = ""
    index: int
    for index in range(0, num_lines):
        block_of_text += "line{}\n".format(index)
    return block_of_text
