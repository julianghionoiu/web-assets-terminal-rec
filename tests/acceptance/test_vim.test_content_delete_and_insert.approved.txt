{"version": 2, "width": 60, "height": 21}
[0.0, "i", "~~~~~~~~~~~~~  display_content()  ~~~~~~~~~~~~~"]
[0.5, "o", ""]
[0.5, "i", "## -- save screen status - will be used on exit"]
[0.5, "o", "\u001b[?1049h"]
[0.5, "i", "## -- clear screen and hide cursor"]
[0.5, "o", "\u001b[?25l"]
[0.5, "o", "\u001b[1;21r"]
[0.5, "o", "\u001b[2J"]
[0.5, "i", "## -- draw screen"]
[0.5, "o", "\u001b[31mred????\u001b[0m word2\r\n"]
[0.5, "o", "\r\n"]
[0.5, "i", "## -- restore and show cursor"]
[0.5, "o", "\u001b[1;1H"]
[0.5, "o", "\u001b[?25h"]
[0.5, "i", "~~~~~~~~~~~~~  cursor_right(num_cols=3)  ~~~~~~~~~~~~~"]
[0.5, "i", "## -- max_cursor_right=12"]
[0.5, "i", "## -- actual_cursor_right=3"]
[1.0, "o", ""]
[1.0, "o", "\u001b[1;4H"]
[1.0, "i", "## -- cursor_position=(row:0,col:3)"]
[1.0, "i", "~~~~~~~~~~~~~  delete_at_cursor(num_chars=20)  ~~~~~~~~~~~~~"]
[1.5, "o", ""]
[1.5, "i", "## -- clear line and hide cursor"]
[1.5, "o", "\u001b[?25l"]
[1.5, "o", "\u001b[1;1H"]
[1.5, "o", "\u001b[2K"]
[1.5, "i", "## -- draw line"]
[1.5, "o", "\u001b[31mred\u001b[0m"]
[1.5, "i", "## -- restore and show cursor"]
[1.5, "o", "\u001b[1;4H"]
[1.5, "o", "\u001b[?25h"]
[1.5, "i", "~~~~~~~~~~~~~  type_at_cursor(text=word\n)  ~~~~~~~~~~~~~"]
[2.0, "o", ""]
[2.0, "i", "## -- clear line and hide cursor"]
[2.0, "o", "\u001b[?25l"]
[2.0, "o", "\u001b[1;1H"]
[2.0, "o", "\u001b[2K"]
[2.0, "i", "## -- draw line"]
[2.0, "o", "\u001b[31mred\u001b[0mw"]
[2.0, "i", "## -- restore and show cursor"]
[2.0, "o", "\u001b[1;4H"]
[2.0, "o", "\u001b[?25h"]
[2.1, "o", ""]
[2.1, "i", "## -- clear line and hide cursor"]
[2.1, "o", "\u001b[?25l"]
[2.1, "o", "\u001b[1;1H"]
[2.1, "o", "\u001b[2K"]
[2.1, "i", "## -- draw line"]
[2.1, "o", "\u001b[31mred\u001b[0mwo"]
[2.1, "i", "## -- restore and show cursor"]
[2.1, "o", "\u001b[1;5H"]
[2.1, "o", "\u001b[?25h"]
[2.2, "o", ""]
[2.2, "i", "## -- clear line and hide cursor"]
[2.2, "o", "\u001b[?25l"]
[2.2, "o", "\u001b[1;1H"]
[2.2, "o", "\u001b[2K"]
[2.2, "i", "## -- draw line"]
[2.2, "o", "\u001b[31mred\u001b[0mwor"]
[2.2, "i", "## -- restore and show cursor"]
[2.2, "o", "\u001b[1;6H"]
[2.2, "o", "\u001b[?25h"]
[2.3, "o", ""]
[2.3, "i", "## -- clear line and hide cursor"]
[2.3, "o", "\u001b[?25l"]
[2.3, "o", "\u001b[1;1H"]
[2.3, "o", "\u001b[2K"]
[2.3, "i", "## -- draw line"]
[2.3, "o", "\u001b[31mred\u001b[0mword"]
[2.3, "i", "## -- restore and show cursor"]
[2.3, "o", "\u001b[1;7H"]
[2.3, "o", "\u001b[?25h"]
[2.4, "o", ""]
[2.4, "i", "## -- clear line and hide cursor"]
[2.4, "o", "\u001b[?25l"]
[2.4, "o", "\u001b[1;1H"]
[2.4, "o", "\u001b[2K"]
[2.4, "i", "## -- draw line"]
[2.4, "o", "\u001b[31mred\u001b[0mword\n"]
[2.4, "i", "## -- restore and show cursor"]
[2.4, "o", "\u001b[1;8H"]
[2.4, "o", "\u001b[?25h"]
[2.5, "o", ""]
